File: examples/node/oauth_web_flow/americommerce.js
var https = require('https'),
		crypto = require('crypto'),
		util = require('util'),
		config = require('./config');

function createSignature() {
	var args = Array.prototype.slice.call(arguments),
			combined = args.join(""),
			sha256 = crypto.createHash('sha256');

	return sha256.update(combined, 'utf8').digest('hex');
}

exports.startNegotiation = function(returnUrl) {
	return util.format("https://%s/api/oauth?app_id=%d&scope=%s&redirect_url=%s",
		config.storeDomain, config.appId, config.appScope, returnUrl);
};

exports.getToken = function(tempData, callback) {
	var sig = createSignature(config.appSecret, tempData.code, config.appId, config.appScope, tempData.returnUrl.toLowerCase());
	var outgoingData = JSON.stringify({
		'app_id': config.appId,
		'auth_id': tempData.authId,
		'signature': sig
	});
	var requestOptions = {
		host: config.storeDomain,
		port: 443,
		path: '/api/oauth/access_token',
		method: 'POST',
		headers: {
			'Content-Type': 'application/json',
			'Content-Length': Buffer.byteLength(outgoingData, 'utf8')
		},
		rejectUnauthorized: false
	};
	var req = https.request(requestOptions, function(res) {
		var data = '';

		res.on('data', function(chunk) {
			data += chunk;
		});
		res.on('end', function() {
			var d = JSON.parse(data);
			callback && callback(d);
		});
	});
	req.write(outgoingData);
	req.end();
};

exports.getProductList = function(token, page, callback) {
	var requestOptions = {
		host: config.storeDomain,
		port: 443,
		path: util.format("/api/v1/products?page=%d", page),
		method: 'GET',
		headers: {
			'X-AC-Auth-Token': token
		},
		rejectUnauthorized: false
	};
	var req = https.request(requestOptions, function(res) {
		var data = '';

		res.on('data', function(chunk) {
			data += chunk;
		});
		res.on('end', function() {
			var d = JSON.parse(data);
			callback && callback(d);
		});
	});
	req.end();
};

File: examples/python/oauth_web_flow/americommerce.py
import requests
import json
import sys
from hashlib import sha256

class ApiClient():

    def __init__(self):
        with open('config.json') as config_file:
            config = json.load(config_file)
        self.app_id = config['app_id']
        self.app_secret = config['app_secret']
        self.app_scope = config['app_scope']
        self.store_domain = config['store_domain']

    def start_negotiation(self, return_url):
        return "https://{}/api/oauth?app_id={}&scope={}&redirect_url={}".format(
            self.store_domain, self.app_id, self.app_scope, return_url)

    def verify(self, return_url, auth_id, code):
        sig = self._create_signature(self.app_secret, code, self.app_id, self.app_scope, return_url.lower())
        headers = {
            'Content-Type': 'application/json'
        }
        body = {
            'app_id': self.app_id,
            'auth_id': auth_id,
            'signature': sig
        }
        uri = "https://{}/api/oauth/access_token".format(self.store_domain)
        # include verify=False if using dev certificate
        # r = requests.post(uri, headers=headers, data=json.dumps(body), verify=False)
        r = requests.post(uri, headers=headers, data=json.dumps(body))
        return r.json()

    def get_product_list(self, token, page = 1):
        headers = {
            'X-AC-Auth-Token': token
        }
        uri = "https://{}/api/v1/products?page={}".format(self.store_domain, page)
        # include verify=False if using dev certificate
        # r = requests.get(uri, headers=headers, verify=False)
        r = requests.get(uri, headers=headers)
        return r.json()

    def _create_signature(self, *args):
        combined = "".join([str(s) for s in args]).encode(encoding='utf-8')
        return sha256(combined).hexdigest()